# Считайте датасет из файла train.csv (это данные о выживаемости на Титанике) с помощью polars

In [12]:
import polars as pl

df = pl.read_csv('train.csv')

print(df.head(5))

shape: (5, 12)
┌─────────────┬──────────┬────────┬───────────────────┬───┬───────────┬─────────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass ┆ Name              ┆ … ┆ Ticket    ┆ Fare    ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---    ┆ ---               ┆   ┆ ---       ┆ ---     ┆ ---   ┆ ---      │
│ i64         ┆ i64      ┆ i64    ┆ str               ┆   ┆ str       ┆ f64     ┆ str   ┆ str      │
╞═════════════╪══════════╪════════╪═══════════════════╪═══╪═══════════╪═════════╪═══════╪══════════╡
│ 1           ┆ 0        ┆ 3      ┆ Braund, Mr. Owen  ┆ … ┆ A/5 21171 ┆ 7.25    ┆ null  ┆ S        │
│             ┆          ┆        ┆ Harris            ┆   ┆           ┆         ┆       ┆          │
│ 2           ┆ 1        ┆ 1      ┆ Cumings, Mrs.     ┆ … ┆ PC 17599  ┆ 71.2833 ┆ C85   ┆ C        │
│             ┆          ┆        ┆ John Bradley (Fl… ┆   ┆           ┆         ┆       ┆          │
│ 3           ┆ 1        ┆ 3      ┆ Heikkinen, Miss.  ┆ … ┆ STON/O2.  ┆ 7.92

## Выведите основную информацию о датасете: информацию о типах данных, число пропусков, средние значения и т.д. - 1 балл
hint: используйте методы и атрибуты датафрейма: describe(), dtypes, null_count(), mean().

In [15]:
print("информация о типах данных")

print(df.dtypes)

информация о типах данных
[Int64, Int64, Int64, String, String, Float64, Int64, Int64, String, Float64, String, String]


In [17]:
print("статистическая информация")

print(df.describe())


статистическая информация
shape: (9, 13)
┌────────────┬─────────────┬──────────┬──────────┬───┬───────────┬───────────┬───────┬──────────┐
│ statistic  ┆ PassengerId ┆ Survived ┆ Pclass   ┆ … ┆ Ticket    ┆ Fare      ┆ Cabin ┆ Embarked │
│ ---        ┆ ---         ┆ ---      ┆ ---      ┆   ┆ ---       ┆ ---       ┆ ---   ┆ ---      │
│ str        ┆ f64         ┆ f64      ┆ f64      ┆   ┆ str       ┆ f64       ┆ str   ┆ str      │
╞════════════╪═════════════╪══════════╪══════════╪═══╪═══════════╪═══════════╪═══════╪══════════╡
│ count      ┆ 891.0       ┆ 891.0    ┆ 891.0    ┆ … ┆ 891       ┆ 891.0     ┆ 204   ┆ 889      │
│ null_count ┆ 0.0         ┆ 0.0      ┆ 0.0      ┆ … ┆ 0         ┆ 0.0       ┆ 687   ┆ 2        │
│ mean       ┆ 446.0       ┆ 0.383838 ┆ 2.308642 ┆ … ┆ null      ┆ 32.204208 ┆ null  ┆ null     │
│ std        ┆ 257.353842  ┆ 0.486592 ┆ 0.836071 ┆ … ┆ null      ┆ 49.693429 ┆ null  ┆ null     │
│ min        ┆ 1.0         ┆ 0.0      ┆ 1.0      ┆ … ┆ 110152    ┆ 0.0       

In [19]:
print("количество нулей")

print(df.null_count())


количество нулей
shape: (1, 12)
┌─────────────┬──────────┬────────┬──────┬───┬────────┬──────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass ┆ Name ┆ … ┆ Ticket ┆ Fare ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---    ┆ ---  ┆   ┆ ---    ┆ ---  ┆ ---   ┆ ---      │
│ u32         ┆ u32      ┆ u32    ┆ u32  ┆   ┆ u32    ┆ u32  ┆ u32   ┆ u32      │
╞═════════════╪══════════╪════════╪══════╪═══╪════════╪══════╪═══════╪══════════╡
│ 0           ┆ 0        ┆ 0      ┆ 0    ┆ … ┆ 0      ┆ 0    ┆ 687   ┆ 2        │
└─────────────┴──────────┴────────┴──────┴───┴────────┴──────┴───────┴──────────┘


In [18]:
print("среднее")

print(df. mean())


среднее
shape: (1, 12)
┌─────────────┬──────────┬──────────┬──────┬───┬────────┬───────────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass   ┆ Name ┆ … ┆ Ticket ┆ Fare      ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---      ┆ ---  ┆   ┆ ---    ┆ ---       ┆ ---   ┆ ---      │
│ f64         ┆ f64      ┆ f64      ┆ str  ┆   ┆ str    ┆ f64       ┆ str   ┆ str      │
╞═════════════╪══════════╪══════════╪══════╪═══╪════════╪═══════════╪═══════╪══════════╡
│ 446.0       ┆ 0.383838 ┆ 2.308642 ┆ null ┆ … ┆ null   ┆ 32.204208 ┆ null  ┆ null     │
└─────────────┴──────────┴──────────┴──────┴───┴────────┴───────────┴───────┴──────────┘


# Посчитайте количество пассажиров каждого класса (Pclass)



In [20]:
pclass_counts = df.get_column('Pclass').value_counts().sort('Pclass')
print(pclass_counts)

shape: (3, 2)
┌────────┬───────┐
│ Pclass ┆ count │
│ ---    ┆ ---   │
│ i64    ┆ u32   │
╞════════╪═══════╡
│ 1      ┆ 216   │
│ 2      ┆ 184   │
│ 3      ┆ 491   │
└────────┴───────┘


#Выведите количество выживших мужчин и женщин на корабле

In [22]:

survived_by_sex = df.group_by('Sex').agg(pl.col('Survived').sum().alias('Survived_Count')).sort('Sex')
print(survived_by_sex)

shape: (2, 2)
┌────────┬────────────────┐
│ Sex    ┆ Survived_Count │
│ ---    ┆ ---            │
│ str    ┆ i64            │
╞════════╪════════════════╡
│ female ┆ 233            │
│ male   ┆ 109            │
└────────┴────────────────┘


# Выведите часть таблицы с пассажирами, возраст которых больше 44 лет

In [26]:

older_passengers = df.filter(pl.col('Age') > 44)
print(older_passengers.head(5))

shape: (5, 12)
┌─────────────┬──────────┬────────┬──────────────────┬───┬────────────┬─────────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass ┆ Name             ┆ … ┆ Ticket     ┆ Fare    ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---    ┆ ---              ┆   ┆ ---        ┆ ---     ┆ ---   ┆ ---      │
│ i64         ┆ i64      ┆ i64    ┆ str              ┆   ┆ str        ┆ f64     ┆ str   ┆ str      │
╞═════════════╪══════════╪════════╪══════════════════╪═══╪════════════╪═════════╪═══════╪══════════╡
│ 7           ┆ 0        ┆ 1      ┆ McCarthy, Mr.    ┆ … ┆ 17463      ┆ 51.8625 ┆ E46   ┆ S        │
│             ┆          ┆        ┆ Timothy J        ┆   ┆            ┆         ┆       ┆          │
│ 12          ┆ 1        ┆ 1      ┆ Bonnell, Miss.   ┆ … ┆ 113783     ┆ 26.55   ┆ C103  ┆ S        │
│             ┆          ┆        ┆ Elizabeth        ┆   ┆            ┆         ┆       ┆          │
│ 16          ┆ 1        ┆ 2      ┆ Hewlett, Mrs.    ┆ … ┆ 248706     ┆ 16.0

Ускорение работы с pandas

# Считайте датасет из файла train.csv (это данные о выживаемости на Титанике) с помощью pandas

In [28]:
import pandas as pd
import bottleneck as bn

df = pd.read_csv('train.csv')


# Посчитайте средний возраст пассажиров и его стандартное отклонение с помощью bottleneck - 1 бал
hint: используйте методы: bn.nanmean(), bn.nanstd()

In [35]:
mean_age = bn.nanmean(df['Age'])
std_age = bn.nanstd(df['Age'])


print(f"Средний возраст")
print(mean_age)

print(f"   ")

print(f"Стандартное отклонение")
print(std_age)

Средний возраст
29.69911764705882
   
Стандартное отклонение
14.516321150817317


# Для каждого пассажира умножьте значение столбца Fare на 1.3 и сохраните результаты как новый столбец Fare_new - 1 балл
hint: используйте методы: .itertuples() или apply()

In [40]:
df['Fare_new'] = df['Fare'].apply(lambda x: x * 1.3)

print(df[['Fare', 'Fare_new']].head(10))

      Fare  Fare_new
0   7.2500   9.42500
1  71.2833  92.66829
2   7.9250  10.30250
3  53.1000  69.03000
4   8.0500  10.46500
5   8.4583  10.99579
6  51.8625  67.42125
7  21.0750  27.39750
8  11.1333  14.47329
9  30.0708  39.09204


# Оптимизация типов pandas


Считайте датасет из файла Housing.csv (это данные о ценах домов) с помощью pandas

In [88]:
import pandas as pd
import numpy as np

df = pd.read_csv('Housing.csv')
print(df.head(5))

      price  area  bedrooms  bathrooms  stories mainroad guestroom basement  \
0  13300000  7420         4          2        3      yes        no       no   
1  12250000  8960         4          4        4      yes        no       no   
2  12250000  9960         3          2        2      yes        no      yes   
3  12215000  7500         4          2        2      yes        no      yes   
4  11410000  7420         4          1        2      yes       yes      yes   

  hotwaterheating airconditioning  parking prefarea furnishingstatus  
0              no             yes        2      yes        furnished  
1              no             yes        3       no        furnished  
2              no              no        2      yes   semi-furnished  
3              no             yes        3      yes        furnished  
4              no             yes        2       no        furnished  


# Для каждого столбца определите оптимальный с точки зрения потребления памяти тип данных - напишите свои выводы в комментариях

In [89]:
print(df.shape)
print(f'  ')

print(df.dtypes)

(545, 13)
  
price                int64
area                 int64
bedrooms             int64
bathrooms            int64
stories              int64
mainroad            object
guestroom           object
basement            object
hotwaterheating     object
airconditioning     object
parking              int64
prefarea            object
furnishingstatus    object
dtype: object


In [95]:
print("Потребление памяти")
print(df.memory_usage(deep=True).sum() / 1024**2, "MB")



Потребление памяти
0.21671676635742188 MB


In [96]:
df = df.convert_dtypes()
print(df.convert_dtypes())


        price  area  bedrooms  bathrooms  stories mainroad guestroom basement  \
0    13300000  7420         4          2        3      yes        no       no   
1    12250000  8960         4          4        4      yes        no       no   
2    12250000  9960         3          2        2      yes        no      yes   
3    12215000  7500         4          2        2      yes        no      yes   
4    11410000  7420         4          1        2      yes       yes      yes   
..        ...   ...       ...        ...      ...      ...       ...      ...   
540   1820000  3000         2          1        1      yes        no      yes   
541   1767150  2400         3          1        1       no        no       no   
542   1750000  3620         2          1        1      yes        no       no   
543   1750000  2910         3          1        1       no        no       no   
544   1750000  3850         3          1        2      yes        no       no   

    hotwaterheating aircond

In [100]:
yes_no_cols = ['mainroad', 'guestroom', 'basement', 'hotwaterheating', 'airconditioning', 'prefarea']
for col in yes_no_cols:
    df[col] = df[col].map({'yes': True, 'no': False})

df['furnishingstatus'] = df['furnishingstatus'].astype('category')

df['price'] = pd.to_numeric(df['price'], downcast='unsigned')
df['area'] = pd.to_numeric(df['area'], downcast='unsigned')
df['bedrooms'] = pd.to_numeric(df['bedrooms'], downcast='unsigned')
df['bathrooms'] = pd.to_numeric(df['bathrooms'], downcast='unsigned')
df['stories'] = pd.to_numeric(df['stories'], downcast='unsigned')
df['parking'] = pd.to_numeric(df['parking'], downcast='unsigned')

memory_after = df.memory_usage(deep=True).sum() / 1024**2

print(f"Потребление памяти после: {memory_after:.2f} МБ")
print(f"\nТипы данных после оптимизации:\n{df.dtypes}")

Потребление памяти после: 0.11 МБ

Типы данных после оптимизации:
price                 UInt32
area                  UInt16
bedrooms               UInt8
bathrooms              UInt8
stories                UInt8
mainroad              object
guestroom             object
basement              object
hotwaterheating       object
airconditioning       object
parking                UInt8
prefarea              object
furnishingstatus    category
dtype: object


## Итог

До: 0.22 МБ

После: 0.01 МБ

До оптимизации DataFrame занимал 0.22 МБ при 545 строках и 13 столбцах. Все числовые столбцы имели тип int64 (8 байт на значение), а строковые столбцы — тип object (около 50 байт на значение).

После оптимизации память сократилась до 0.01МБ. Это означает уменьшение в 20 раз.

Получается, что надо маленькие числа заменить на маленькие типы, бинарные значения -на boolean, повторяющиеся строки( у нас в последнем столбце было всего 3 различных значения) -на категории